### PR TITLE
Task crud

### PR DESCRIPTION
**Pull request by tinoquang**

### CODE REVIEW

File: pkg/api/task.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	GetCompletedTasks(ctx context.Context) ([]db.Task, error)
+	GetIncompletedTasks(ctx context.Context) ([]db.Task, error)
+	Update(ctx context.Context, id int, taskName string, completed bool) (db.Task, error)
+	Delete(ctx context.Context, id int) error
+}
+
+// Task in body request
+func (h *APIHandler) CreateTask(c echo.Context) error {
+
+	taskName := c.FormValue("task")
```
Comment (vdhieu) [inline pkg/api/task.go:27]: use validator for create/update request
https://echo.labstack.com/guide/request

File: pkg/api/task.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+			return c.String(http.StatusNotFound, "Task not found")
+		}
+		logging.Danger(err)
+		return c.String(http.StatusInternalServerError, "Internal Server Error")
+	}
+
+	return c.JSON(http.StatusOK, task)
+}
+
+func (h *APIHandler) Tasks(c echo.Context) error {
```
Comment (vdhieu) [inline pkg/api/task.go:60]: Too short function naming, should be `GetTask` or `GetTaskList`

File: pkg/api/task.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+	status := c.QueryParam("status")
```
Comment (vdhieu) [inline pkg/api/task.go:62]: should have pagination

File: pkg/api/task.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+		return c.String(http.StatusBadRequest, "Invalid ID number")
+	}
+
+	task, err := h.taskStore.Get(c.Request().Context(), id)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return c.String(http.StatusNotFound, "Task not found")
+		}
+		logging.Danger(err)
+		return c.String(http.StatusInternalServerError, "Internal Server Error")
```
Comment (vdhieu) [inline pkg/api/task.go:54]: API should consistently return 1 data format event for error, in this case, is JSON.

```json
{
    "error": "not found / internal server"
}
```
or 

```json
{
    "message": "not found / internal server"
}
```

File: pkg/api/task.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	_, err = h.taskStore.Update(c.Request().Context(), id, taskName, completed)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return c.String(http.StatusNotFound, "Task not found")
+		}
+		logging.Danger(err)
+		return c.String(http.StatusInternalServerError, "Internal Server Error")
+	}
+
+	return c.String(http.StatusOK, "Task updated")
```
Comment (vdhieu) [inline pkg/api/task.go:105]: update API should return an updated object, not string updated task

File: pkg/api/task.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+		return c.String(http.StatusInternalServerError, "Internal Server Error")
+	}
+
+	err = h.taskStore.Delete(c.Request().Context(), id)
+	if err != nil {
+		logging.Danger(err)
+		return c.String(http.StatusInternalServerError, "Internal Server Error")
+	}
+
+	return c.String(http.StatusOK, "Task deleted")
```
Comment (vdhieu) [inline pkg/api/task.go:131]: we can return deleted object or object id in json format

File: main.go [hunk 7]
```diff
@@ -1,18 +1,33 @@
...
+	_ "github.com/lib/pq"
+	"github.com/tinoquang/todo/pkg/api"
+	"github.com/tinoquang/todo/pkg/store"
+
 	"github.com/labstack/echo/v4"
 )
 
 func main() {
 
+	db, err := sql.Open("postgres", "postgresql://tino:quangtino123@localhost:5432/todo?sslmode=disable")
```
Comment (vdhieu) [inline main.go:16]: read this from env not hard coded

**Review by vdhieu — COMMENTED** [review]
Good work. 
I have few comments
- does sqlc supports separate each model a file? although just a small app but put everything to a single models.go would hard to maintain 🤔
- you forgot to use json tag for the return object => this will return `PascalCase` json object. it's ok however the community is prefer `camelCase` or `snake_case`

**Comment by tinoquang** [comment]
> Good work.
> I have few comments
> 
> * does sqlc supports separate each model a file? although just a small app but put everything to a single models.go would hard to maintain thinking
> * you forgot to use json tag for the return object => this will return `PascalCase` json object. it's ok however the community is prefer `camelCase` or `snake_case`



> Good work.
> I have few comments
> 
> * does sqlc supports separate each model a file? although just a small app but put everything to a single models.go would hard to maintain thinking
> * you forgot to use json tag for the return object => this will return `PascalCase` json object. it's ok however the community is prefer `camelCase` or `snake_case`

As far as I know, sqlc now auto wrap all models in one file. But this file is generated from our migration files so I think we can still manage the models by checking these files

**Pull request merged**